[X86] Add CPU detection for more znver2 CPUs

boomanaiden154 opened the pull request:
This patch adds proper detection support for more znver2 CPUs.

Specifically, this adds in support for CPUs codenamed Renoir, Lucienne,
and Mendocino.

This was originally proposedfor Renoir  in https://reviews.llvm.org/D96220 and
got approved, but slipped through the cracks. However, there is still a
demand for this feature.

Fixes https://github.com/llvm/llvm-project/issues/74934.

boomanaiden154 committed:
[X86] Add detection for more znver2 CPUs

This patch adds proper detection support for more znver2 CPUs.

Specifically, this adds in support for CPUs codenamed Renoir, Lucienne,
and Mendocino.

This was originally proposedfor Renoir  in https://reviews.llvm.org/D96220 and
got approved, but slipped through the cracks. However, there is still a
demand for this feature.

Fixes https://github.com/llvm/llvm-project/issues/74934.

boomanaiden154 commented:
> I tend to double check with https://github.com/torvalds/linux/blob/master/drivers/hwmon/k10temp.c#L440 as that has good Ryzen cpu model coverage

Thanks for the suggestion. https://en.wikichip.org/wiki/amd/cpuid also seems to have some good info. Currently we're covering Rome and Castle Peak (`0x31`), and Matisse (`0x71`). We're missing the Xbox series X APU (`0x47`), Renoir (this patch, `0x60`), Lucienne (`0x68`), Matisse (`0x71`), Van Gogh (`0x90`), and Mendocino (`0xA0`).

I'm assuming we just want to cover all of these?

boomanaiden154 commented:
Trying to find this in the architecture manuals, and I'm not getting very far. There are only two revision guides for 17h. One covers Rome only, and the other covers 1st and 2nd gen Ryzen.

boomanaiden154 commented:
I've updated it to match the `k10temp.c` listing (Mendocino is Zen2 despite using a different core chiplet die offset for the temperature reporting in the kernel from what I can gather). This adds in the following arches:
1. Renoir
2. Lucienne
3. Mendocino

These are all covered by `k10temp.c` (unlike Van Gogh (Steam deck APU) and the Xbox Series X APU which we wouldn't expect to have public information about either way).

I've also adjusted the detection for Rome/Castle Peak to be a specific value rather than a range as only `0x31` seems to exist (and only `0x31` is covered in `k10temp.c`). I've mirrored the range from `k10temp.c` for Mendocino.

boomanaiden154 committed:
Remove moved comment

RKSimon reviewed (dismissed):
LGTM - but please give @ganeshgit another chance to respond before committing in a couple of days

ganeshgit commented:
@RKSimon Thank for bringing it to my notice.
Here is the list!

Family 17h Models 10h-1Fh (Raven1)				Zen
Family 17h Models 10h-1Fh (Picasso)			Zen +
Family 17h Models 20h-2Fh (Raven2 x86)			Zen
Family 17h Models 30h-3Fh (Starship)			Zen 2
Family 17h Models 47h (Cardinal)				Zen 2
Family 17h Models 60h-67h (Renoir)				Zen 2
Family 17h Models 68h-6Fh (Lucienne)			Zen 2
Family 17h Models 70h-7Fh (Matisse)			Zen 2
Family 17h Models 84h-87h (ProjectX)			Zen 2
Family 17h Models 90h-97h (VanGogh)				Zen 2
Family 17h Models 98h-9Fh (Mero)				Zen 2
Family 17h Models A0h-AFh (Mendocino)			Zen 2
Family 19h Models 00h-0Fh (Genesis, Chagall)		Zen 3
Family 19h Models 10h-1Fh (Stones; Storm Peak)		Zen 4
Family 19h Models 20h-2Fh (Vermeer)			Zen 3
Family 19h Models 30h-3Fh (Badami)				Zen 3
Family 19h Models 40h-4Fh (Rembrandt)			Zen 3+
Family 19h Models 50h-5Fh (Cezanne)			Zen 3
Family 19h Models 60h-6Fh (Raphael)			Zen 4
Family 19h Models 70h-77h (Phoenix, Hawkpoint1)	Zen 4
Family 19h Models 78h-7Fh (Phoenix 2, Hawkpoint2)	Zen 4
Family 19h Models A0h-AFh (Stones-Dense)			Zen 4

boomanaiden154 committed:
Update based on reviewer feedback

RKSimon reviewed (approved):
LGTM - please can add equivalent comments for the case 25 zen3/zen4 comments so we don't lose the info that @ganeshgit listed.

boomanaiden154 committed:
Add additional comments

boomanaiden154 committed:
Make zenv1 consistent with the rest of the file

boomanaiden154 commented:
> LGTM - please can add equivalent comments for the case 25 zen3/zen4 comments so we don't lose the info that @ganeshgit listed.

Updated. I also updated the checks to match the comments. In some cases, that resulted in more conditionals, but they are trivially optimizable and make the code more readable (in my opinion). Please let me know if there are any issues with that approach and I can get it fixed.

mikaelholmen commented on compiler-rt/lib/builtins/cpu_model/x86.c:
```diff
@@ -647,35 +647,59 @@ static const char *getAMDProcessorTypeAndSubtype(unsigned Family,
-      // Family 19h Models 30h-3Fh - Zen3
-      // Family 19h Models 40h-4Fh - Zen3+
-      // Family 19h Models 50h-5Fh - Zen3+
+    if ((Model >= 0x00 && Model <= 0x0f) || (Model >= 0x20 && Model <= 0x2f) ||
```
Since ```Model``` is unsigned ```Model >= 0x00``` is always true.
gcc even warns about it
```
../lib/TargetParser/Host.cpp:1163:16: warning: comparison of unsigned expression >= 0 is always true [-Wtype-limits]
 1163 |     if ((Model >= 0x00 && Model <= 0x0f) || (Model >= 0x20 && Model <= 0x2f) ||
      |          ~~~~~~^~~~~~~
```

boomanaiden154 replied:
I think I'd still make the argument that representing all of the ranges consistently makes the code more readable, but I've posted 3ddf3685248a8c3ef07bf8290196c33521edb894 as it doesn't significantly impact readability and gets rid of the gcc warning. Thanks for pointing this issue out!
